codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedIntegerGenerator.kt

```diff
@@ -0,0 +1,187 @@
+
+        val condition = if (rangeTrait.min.isPresent && rangeTrait.max.isPresent) {
+            "(${rangeTrait.min.get()}..=${rangeTrait.max.get()}).contains(&value)"
+        } else if (rangeTrait.min.isPresent) {
+            "${rangeTrait.min.get()} <= value"
+        } else {
+            "value <= ${rangeTrait.max.get()}"
+        }
+
+        val constrainedTypeVisibility = if (publicConstrainedTypes) {
+            Visibility.PUBLIC
+        } else {
+            Visibility.PUBCRATE
+        }
+        val constrainedTypeMetadata = RustMetadata(
+            Attribute.Derives(setOf(RuntimeType.Debug, RuntimeType.Clone, RuntimeType.PartialEq, RuntimeType.Eq, RuntimeType.Hash)),
+            visibility = constrainedTypeVisibility,
+        )
+
+        writer.documentShape(shape, model, note = rustDocsNote(name))
+        constrainedTypeMetadata.render(writer)
+        writer.rust("struct $name(pub(crate) $inner);")
+        if (constrainedTypeVisibility == Visibility.PUBCRATE) {
+            Attribute.AllowUnused.render(writer)
+        }
+        writer.rustTemplate(
+            """
+            impl $name {
+                /// ${rustDocsInnerMethod(inner)}
+                pub fn inner(&self) -> &$inner {
+                    &self.0
+                }
+
+                /// ${rustDocsIntoInnerMethod(inner)}
+                pub fn into_inner(self) -> $inner {
+                    self.0
+                }
+            }
+
+            impl #{ConstrainedTrait} for $name  {
+                type Unconstrained = $inner;
+            }
+
+            impl #{From}<$inner> for #{MaybeConstrained} {
+                fn from(value: $inner) -> Self {
+                    Self::Unconstrained(value)
+                }
+            }
+
+            impl #{From}<$name> for #{MaybeConstrained} {
+                fn from(value: $name) -> Self {
+                    Self::Constrained(value)
+                }
+            }
+
+            impl #{Display} for $name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                   ${shape.redactIfNecessary(model, "self.0")}.fmt(f)
+                }
+            }
+
+            impl #{TryFrom}<$inner> for $name {
+                type Error = #{ConstraintViolation};
+
+                /// ${rustDocsTryFromMethod(name, inner)}
+                fn try_from(value: $inner) -> Result<Self, Self::Error> {
+                    if $condition {
+                        Ok(Self(value))
+                    } else {
+                        Err(#{ConstraintViolation}::Size(value))
+                    }
+                }
+            }
+
+            impl #{From}<$name> for $inner {
+                fn from(value: $name) -> Self {
+                    value.into_inner()
+                }
+            }
+
+            impl #{AsRef}<$inner> for $name {
+                fn as_ref(&self) -> &$inner {
+                    self.inner()
+                }
+            }
```

Comment:
I'd remove this one for now. We don't need it for the generated code, and it's unclear what traits we should implement for newtypes.

Reply:
Done in [0b80d2e](https://github.com/awslabs/smithy-rs/pull/2005/commits/0b80d2e9c2fd13b1a469b07d2b1074aec79469de) (and https://github.com/awslabs/smithy-rs/pull/2005/commits/58c156fd0e1d95caeb4fc7f1b7efd6c98889003f)